test/e2e/apps/job.go

```diff
@@ -112,21 +114,45 @@ var _ = SIGDescribe("Job", func() {
 	// Pods sometimes fail, but eventually succeed, after pod restarts
 	ginkgo.It("should run a job to completion when tasks sometimes fail and are not locally restarted", func() {
 		ginkgo.By("Creating a job")
-		// 50% chance of container success, local restarts.
+		// Success and failure alternate, no local restarts.
 		// Can't use the failOnce approach because that relies
 		// on an emptyDir, which is not preserved across new pods.
-		// Worst case analysis: 15 failures, each taking 1 minute to
-		// run due to some slowness, 1 in 2^15 chance of happening,
-		// causing test flake.  Should be very rare.
-		// With the introduction of backoff limit and high failure rate this
-		// is hitting its timeout, the 3 is a reasonable that should make this
-		// test less flaky, for now.
-		job := jobutil.NewTestJob("randomlySucceedOrFail", "rand-non-local", v1.RestartPolicyNever, parallelism, 3, nil, 999)
+		// To avoid flakes, make the job pods wait for exit code
+		// and use an exec pod to command them succeed and fail
+		// alternately.
+		job := jobutil.NewTestJob("waitForExitCode", "rand-non-local", v1.RestartPolicyNever, parallelism, 3, nil, 999)
 		job, err := jobutil.CreateJob(f.ClientSet, f.Namespace.Name, job)
 		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
 
+		execPod := e2epod.CreateExecPodOrFail(f.ClientSet, f.Namespace.Name, "execpod", nil)
 		ginkgo.By("Ensuring job reaches completions")
-		err = jobutil.WaitForJobComplete(f.ClientSet, f.Namespace.Name, job.Name, *job.Spec.Completions)
+		numPods := 0
+		err = wait.PollImmediate(framework.Poll, jobutil.JobTimeout, func() (bool, error) {
+			curr, err := jobutil.GetJob(f.ClientSet, f.Namespace.Name, job.Name)
+			if err != nil {
+				return false, err
+			}
+			for _, c := range curr.Status.Conditions {
+				if c.Type == batchv1.JobComplete && c.Status == v1.ConditionTrue {
+					return true, nil
+				} else if c.Type == batchv1.JobFailed && c.Status == v1.ConditionTrue {
+					return false, fmt.Errorf("job %s has failed", job.Name)
+				}
+			}
+			pods, err := jobutil.GetJobPods(f.ClientSet, f.Namespace.Name, job.Name)
+			if err != nil {
+				return false, fmt.Errorf("failed to get PodList for job %s in namespace: %s", job.Name, f.Namespace.Name)
+			}
+			for _, pod := range pods.Items {
+				if pod.Status.Phase == v1.PodRunning {
+					// Fail if numPods is odd, otherwise succeed.
+					cmd := fmt.Sprintf("curl -q -s --connect-timeout 2 http://%s/exit?code=%d", pod.Status.PodIP, numPods%2)
```

Comment:
I don't think host curl invocations will work reliably across network topologies and node types. if the hostPath volume ends up not working out, and we need to do a network callback, I'd recommend using the `/api/v1/namespaces/$ns/pods/$pod:80/proxy/exit?code=%d` approach

Reply:
> Fail if numPods is odd, otherwise succeed

it's not clear why this is correct

Reply:
The command was run in a agnhost pod so should work, but anyway I have changed to use hostPath volume and schedule Pods to a single Node as you suggested.